src/jdk.jlink/share/classes/jdk/tools/jlink/internal/plugins/SystemModulesPlugin.java

```diff
@@ -1266,20 +1231,42 @@ void newRequires(Set<Requires.Modifier> mods, String name, String compiledVersio
              *
              */
             void exports(Set<Exports> exports) {
-                cob.aload(BUILDER_VAR)
-                   .loadConstant(exports.size())
-                   .anewarray(CD_EXPORTS);
+                cob.aload(BUILDER_VAR);
+                loadExportsArray(exports);
+                cob.invokevirtual(CD_MODULE_BUILDER,
+                        "exports",
+                        MTD_EXPORTS_ARRAY)
+                        .pop();
+            }
+
+            void loadExportsArray(Set<Exports> exports) {
+                if (exports.size() > SET_SIZE_THRESHOLD) {
+                    String methodName = "module" + index + "Exports";
+                    addModuleHelpers(clb -> clb.withMethodBody(
+                            methodName,
+                            MethodTypeDesc.of(CD_EXPORTS.arrayType()),
+                            ACC_PRIVATE | ACC_FINAL | ACC_STATIC,
```

Comment:
Just curious, why do we mark our split provider methods final? The final flag is only used for hiding.

Reply:
Thanks for pointing that out, I just follow language convention and did not think of that.

Reply:
In language convention, I don't think we declare our private static utility methods as final either.